// @progress/kendo-angular-popup/dist/npm/services/scrollable.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var fromEvent_1 = require("rxjs/observable/fromEvent");
var merge_1 = require("rxjs/observable/merge");
var auditTime_1 = require("rxjs/operators/auditTime");
var dom_service_1 = require("./dom.service");
var util_1 = require("../util");
/**
 * @hidden
 */
exports.THRESHOLD_DIFF = 1;
/**
 * @hidden
 */
var ScrollableService = /** @class */ (function () {
    function ScrollableService(_dom, _zone) {
        this._dom = _dom;
        this._zone = _zone;
    }
    ScrollableService.prototype.forElement = function (element) {
        this.unsubscribe();
        this.element = element;
        return this;
    };
    ScrollableService.prototype.subscribe = function (callback) {
        var _this = this;
        if (!callback || !util_1.isDocumentAvailable() || !this.element) {
            return;
        }
        var nativeElement = this._dom.nativeElement(this.element);
        var parents = this._dom.scrollableParents(this.element);
        this._zone.runOutsideAngular(function () {
            var observables = parents.map(function (p) { return fromEvent_1.fromEvent(p, "scroll").pipe(auditTime_1.auditTime(util_1.FRAME_DURATION)); });
            var subscriber = function (e) {
                callback(_this.isVisible(nativeElement, e.target));
            };
            _this.subscription = merge_1.merge.apply(void 0, observables).subscribe(subscriber);
        });
    };
    ScrollableService.prototype.unsubscribe = function () {
        if (!this.subscription) {
            return;
        }
        this.subscription.unsubscribe();
    };
    ScrollableService.prototype.isVisible = function (elem, container) {
        var elemRect = this._dom.boundingOffset(elem);
        var containerRect = this._dom.boundingOffset(container);
        if ((containerRect.top - elemRect.bottom) > exports.THRESHOLD_DIFF) {
            return false;
        }
        if ((containerRect.bottom - elemRect.bottom) < -exports.THRESHOLD_DIFF) {
            return false;
        }
        if ((containerRect.right - elemRect.right) < -exports.THRESHOLD_DIFF) {
            return false;
        }
        if ((containerRect.left - elemRect.left) > exports.THRESHOLD_DIFF) {
            return false;
        }
        return true;
    };
    ScrollableService.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    ScrollableService.ctorParameters = function () { return [
        { type: dom_service_1.DOMService, },
        { type: core_1.NgZone, },
    ]; };
    return ScrollableService;
}());
exports.ScrollableService = ScrollableService;
