// @progress/kendo-angular-buttons/dist/npm/button/button.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var BehaviorSubject_1 = require("rxjs/BehaviorSubject");
var Subject_1 = require("rxjs/Subject");
/**
 * @hidden
 */
var KendoButtonService = /** @class */ (function () {
    function KendoButtonService() {
        this.buttonLookChange = new BehaviorSubject_1.BehaviorSubject('default');
        this.buttonClicked = new Subject_1.Subject();
        this.buttonClicked$ = this.buttonClicked.asObservable();
    }
    KendoButtonService.prototype.click = function (button) {
        this.buttonClicked.next(button);
    };
    KendoButtonService.prototype.setButtonLook = function (look) {
        this.buttonLookChange.next(look);
    };
    KendoButtonService.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    KendoButtonService.ctorParameters = function () { return []; };
    return KendoButtonService;
}());
exports.KendoButtonService = KendoButtonService;
